src/clr-angular/utils/smart-popover/smart-popover-content.ts

```diff
@@ -0,0 +1,134 @@
+
+// https://github.com/angular/angular/issues/20351#issuecomment-344009887
+/** @dynamic */
+@Directive({ selector: '[clrSmartPopoverContent]' })
+export class ClrSmartPopoverContent implements AfterContentChecked, OnDestroy {
+  private view: EmbeddedViewRef<void>;
+  private subscriptions: Subscription[] = [];
+
+  @Input('clrSmartPopoverContent')
+  public set open(value: boolean) {
+    this.smartOpenService.open = !!value;
+  }
+
+  @Input('clrSmartPopoverContentAt')
+  set clrSmartOpenContentAt(position: ClrSmartPosition) {
+    this.smartPositionService.position = position;
+  }
+
+  @Input('clrSmartPopoverContentOutsideClickToClose')
+  set outsideClickClose(clickToClose) {
+    this.smartEventsService.outsideClickClose = !!clickToClose;
+  }
+
+  @Input('clrSmartPopoverContentScrollToClose')
+  set scrollToClose(scrollToClose) {
+    this.smartEventsService.scrollToClose = !!scrollToClose;
+  }
+
+  constructor(
+    @Inject(DOCUMENT) private document: HTMLDocument,
+    private container: ViewContainerRef,
+    private template: TemplateRef<any>,
+    private renderer: Renderer2,
+    private smartPositionService: ClrSmartPopoverPositionService,
+    private smartEventsService: ClrSmartPopoverEventsService,
+    private smartOpenService: ClrSmartPopoverToggleService
+  ) {}
+
+  ngAfterViewInit() {
+    this.subscriptions.push(
+      this.smartOpenService.openChange.subscribe(change => {
+        if (change) {
+          this.addContent();
+        } else {
+          this.removeContent();
+        }
+      })
+    );
+  }
+
+  ngOnDestroy() {
+    this.removeContent();
+    this.subscriptions.forEach(sub => sub.unsubscribe());
+  }
+
+  private removeContent(): void {
+    if (!this.view) {
+      return;
+    }
+    this.view.rootNodes.forEach(node => this.renderer.removeChild(this.document.body, node));
+    this.container.clear();
+    delete this.view;
+
+    // Clean up for when popovers do have to go off screen
+    this.renderer.removeStyle(this.document.body, 'position');
+    this.renderer.removeStyle(this.document.body, 'width');
+    this.renderer.removeStyle(this.document.body, 'height');
+
+    this.hasPositionCoords = false;
+  }
+
+  /**
+   * TODO(matt): investigate why DebugElement retains a reference to the nodes and causes a memory leak.
+   * A note about the use of appendChild/removeChild
+   * The DebugElement is keeping a reference to the detached node and its unclear why.
+   * This does warrant further investigation. But, since it doesn't happen in production mode
+   * it is a low priority issue for now.
+   */
+  private addContent() {
+    // Create the view container
+    this.view = this.container.createEmbeddedView(this.template);
+    this.smartEventsService.contentRef = this.view.rootNodes[0]; // So we know where/what to set close focus on
+
+    // These three lines prevent issues when popover content does have to go offscreen
+    this.renderer.setStyle(this.document.body, 'position', 'fixed');
```

Comment:
Isn't this a bit too global? Don't we risk here to override some fancy stuff that the users of Clarity may have done with their own `body`s?
Same for the next 2 lines.
Also removing these in removeContent() seems a bit dangerous to me, too.

Reply:
This is only added whne content is open. There may be popover content elements that we cannot fix (super tall popovers that go off screen top/bottom or left/right). In that case there is no safe place we can position it to. However because the popover content is outside the Clarity app container the browser will get scroll bars. When I was developing with all of the potential violations, it was no a great user experience.

Reply:
It's quite unlikely that the user will put his own `position` style on the body, but if he did (for example, we do it, so why not someone else decides to do it), then after opening and closing one popup, his setting is gone.

Reply:
In the fix for the firefox bug I was able to remove the need to add any styles on the body. This is no longer a concern.